R/parallelization.R

```diff
@@ -0,0 +1,144 @@
+#' [`daemons()`][mirai::daemons]. This is a function from the \pkg{mirai}
+#' package that sets up daemons (persistent background processes that receive
+#' parallel computations) on your local machine or across the network.
+#'
+#' \pkg{purrr} requires these to be set up prior to performing any parallel map
+#' operations. It is usual to set daemons once per session. You can leave them
+#' running as they consume little in resources whilst waiting to receive tasks.
+#'
+#' If they are not set, you will see a warning from the \pkg{mirai} package that
+#' it is launching one local daemon - this ensures that the function proceeds,
+#' but offers no actual parallelization and is hence probably not what you would
+#' want.
+#'
+#' `daemons()`arguments:
+#'
+#' * `n`: the number of daemons to launch on your local machine, e.g.
+#'   `daemons(7)`. As a rule of thumb, for maximum efficiency this should be (at
+#'   most) one less than the number of cores on your machine, leaving one core
+#'   for the main R process.
+#' * `url` and `remote`: used to set up and launch daemons for distributed
+#'   computing over the network. See [daemons][mirai::daemons] function
+#'   documentation for more details.
+#' * None: calling `daemons()` with no arguments returns a summary of the
+#'   current connection status and mirai tasks queued and completed.
+#'
+#' For more details, including on further arguments, see
+#' [daemons][mirai::daemons].
+#'
+#' Resetting daemons:
+#'
+#' `daemons(0)` resets and terminates any existing daemons.
+#'
+#' All daemons automatically terminate when you end your session and the
+#' connection drops. For this reason, it is not required to specifically
+#' terminate daemons in this instance, although it is still good practice to do
+#' so.
+#'
+#' ## with() method
+#'
+#' The `with()` method for daemons provides a convenient way of evaluating a
+#' statement with daemons set up for the duration of the statement.
+#'
+#' As an example, the below is evaluated in parallel using 7 daemons:
+#'
+#' \preformatted{
+#' with(daemons(7), {
+#'   1:10 |>
+#'   map(rnorm, n = 10, .parallel = TRUE) |>
+#'   map_dbl(mean, .parallel = TRUE)
+#' })
+#' }
+#'
+#' ## Further documentation
+#'
+#' \pkg{purrr}'s parallelization is powered by \CRANpkg{mirai}, so see the
+#' [mirai introduction and reference](https://shikokuchuo.net/mirai/articles/mirai.html)
+#' for more details.
+#'
+#' @name parallelization
+NULL
+
+mmap_ <- function(.x, .f, .args, .parallel, .progress, .type, error_call) {
+
+  m <- if (isTRUE(.parallel)) {
+    mirai::mirai_map(.x, .f, .args = .args)
+  } else if (is.list(.parallel)) {
+    mirai::mirai_map(.x, .f, as.environment(.parallel), .args = .args)
+  } else {
+    cli::cli_abort(
+      "'.parallel' must be TRUE/FALSE or a list, not a {.obj_type_friendly {(.parallel)}}.",
+      call = error_call
+    )
+  }
+
+  options <- c(".stop", if (isTRUE(.progress)) ".progress")
```

Comment:
I'm not entirely sure how you'd expose this as an option, but I would not be surprised if we eventually got requests for a way to opt out of "early stopping", instead running everything to completion/failure.

I think what you have here is probably fine for now, just thinking out loud about this one

Reply:
That would require purr to actually return something other than error. 

It's of course better to stop early as soon as we know we're going to error if that's all we're going to do anyway.

I think this would be another `.` argument as it would apply to purrr in general (it wouldn't be useful just for the parallel case?)  When that happens, it can then be handled the same way as `.progress`.